### PR TITLE
Added python SDK examples

### PR DESCRIPTION
**Pull request by agazeley**

### CODE REVIEW

**Comment by agazeley** [comment]
@puneetsarna

File: snippets/sdk/python/deployments/create_or_update.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        },
+    ).result()
+    print("Created a deployment:\n{}".format(deployment))
+
+    # Get deployment
+    deployment = nginx_client.deployments.get(GROUP_NAME, DEPLOYMENT_NAME)
+    print("Get deployment:\n{}".format(deployment))
+
+    # Update deployment with a new tag
+    deployment = nginx_client.deployments.begin_update(
```
Comment (puneetsarna) [inline snippets/sdk/python/deployments/create_or_update.py:70]: is there a reason to update tags?

File: snippets/sdk/python/deployments/create_or_update.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    # Create clients
+    # For other authentication approaches, please see: https://pypi.org/project/azure-identity/
+    nginx_client = NginxManagementClient(
+        credential=DefaultAzureCredential(), subscription_id=SUBSCRIPTION_ID
+    )
+    # Create prerequisite resources
+    subnet, public_ip, identity = prerequisites.create()
+
+    # - private IP configuration example
+    network_profile_private = {
```
Comment (puneetsarna) [inline snippets/sdk/python/deployments/create_or_update.py:26]: unused var?

File: snippets/sdk/python/deployments/with_logging.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+from azure.mgmt.storage import StorageManagementClient
+
+import prerequisites
+
+
+def main():
+    SUBSCRIPTION_ID = os.environ.get("AZURE_SUBSCRIPTION_ID", None)
+    GROUP_NAME = "myResourceGroup"
+    DEPLOYMENT_NAME = "myDeployment"
+    STORAGE_ACCOUNT_NAME = "mystorageaccountxyzaaa"
```
Comment (puneetsarna) [inline snippets/sdk/python/deployments/with_logging.py:17]: not: if you can align the name

File: snippets/sdk/python/certificates/create_or_update.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                "enabled_for_deployment": True,
+                "enabled_for_disk_encryption": True,
+                "enabled_for_template_deployment": True,
+            },
+        },
+    ).result()
+    print("Created a vault:\n{}".format(vault))
+
+    # Give current user key vault administrator role assignment
+    role_id = "00482a5a-887f-4fb3-b363-3b7fe8e74483"
```
Comment (ryepup) [inline snippets/sdk/python/certificates/create_or_update.py:74]: can we use names instead of magic uuids?

```suggestion
    role_id = "Key Vault Administrator"
```

File: snippets/sdk/python/certificates/create_or_update.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    role_definition_id = "/".join(
+        [
+            "subscriptions",
+            SUBSCRIPTION_ID,
+            "providers/Microsoft.Authorization/roleDefinitions",
+            role_id,
+        ]
+    )
+    role_assignment = "88888888-7000-0000-0000-000000000004"
```
Comment (ryepup) [inline snippets/sdk/python/certificates/create_or_update.py:83]: might be worth a comment about which uuids matter and which are arbitrary.

File: snippets/sdk/python/certificates/create_or_update.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    # Create vault
+    # https://github.com/Azure-Samples/azure-samples-python-management/tree/main/samples/keyvault
+    # A managed identity only requires the 'Key Vault Secrets User' role to give NGINXaaS for Azure
+    # the appropriate permissions to pull certificates from AKV
+    vault = keyvault_client.vaults.begin_create_or_update(
+        GROUP_NAME,
+        VAULT_NAME,
+        {
+            "location": "eastus",
```
Comment (ryepup) [inline snippets/sdk/python/certificates/create_or_update.py:59]: maybe pull out a variable at the top of `main()` for the location?

File: snippets/sdk/python/deployments/with_logging.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    # Create prerequisite resources
+    subnet, public_ip, identity = prerequisites.create()
+
+    # Create storage account
+    # https://github.com/Azure-Samples/azure-samples-python-management/tree/main/samples/storage
+    storage_client.storage_accounts.begin_create(
+        GROUP_NAME,
+        STORAGE_ACCOUNT_NAME,
+        {
+            "sku": {"name": "Standard_GRS"},
```
Comment (puneetsarna) [inline snippets/sdk/python/deployments/with_logging.py:41]: try not to recommend this SKU. GRS might cost the user more. Leverage the simplest/default LRS.

File: snippets/sdk/python/certificates/create_or_update.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            role_id,
+        ]
+    )
+    role_assignment = "88888888-7000-0000-0000-000000000004"
+    auth_client.role_assignments.create(
+        vault.id,
+        role_assignment,
+        {
+            "role_definition_id": role_definition_id,
+            "principal_id": os.getenv("AZURE_CLIENT_ID"),
```
Comment (ryepup) [inline snippets/sdk/python/certificates/create_or_update.py:89]: all the other `getenv`s are at the top of main, maybe put this one there too?

File: snippets/sdk/python/deployments/with_logging.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        role_assignment,
+        {
+            "role_definition_id": role_definition_id,
+            "principal_id": identity.principal_id,
+        },
+    )
+    print("Created a role assignment.\n")
+
+    # Delete a deployment
+    nginx_client.deployments.begin_delete(GROUP_NAME, DEPLOYMENT_NAME).result()
```
Comment (puneetsarna) [inline snippets/sdk/python/deployments/with_logging.py:115]: okay so if I run this, the user may not get enough time to look at what's happening?

Also it now does not align with `create_or_update.py` where you have it commented out? was that intentional?

File: snippets/sdk/python/deployments/with_metrics.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+#! /usr/local/env python3
```
Comment (puneetsarna) [inline snippets/sdk/python/deployments/with_metrics.py:1]: one feedback that I am still confused about myself and not for this MR: do we care about having these different types of examples: one for logging, one for metrics?

Maybe combine them in one and say "enable_diagnostics.py" that enables both?

**Review by puneetsarna — APPROVED** [review]

File: snippets/sdk/python/configurations/archive_file.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+from azure.identity import DefaultAzureCredential
+from azure.mgmt.nginx import NginxManagementClient
+
+
+def main():
+    SUBSCRIPTION_ID = os.environ.get("AZURE_SUBSCRIPTION_ID", None)
+    GROUP_NAME = "myResourceGroup"
+    DEPLOYMENT_NAME = "myDeployment"
+    CONFIGURATION_NAME = "default"  # configuration must be named default
```
Comment (ryepup) [inline snippets/sdk/python/configurations/archive_file.py:14]: does this requirement need to be documented anywhere else?

File: snippets/sdk/python/configurations/archive_file.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    # Get configuration
+    deployment = nginx_client.configurations.get(
+        GROUP_NAME, DEPLOYMENT_NAME, CONFIGURATION_NAME
+    )
+    print("Get configuration:\n{}".format(deployment))
+
+    # Delete our configuration from the deployment
+    nginx_client.configurations.begin_delete(
+        GROUP_NAME, DEPLOYMENT_NAME, CONFIGURATION_NAME
+    )
```
Comment (ryepup) [inline snippets/sdk/python/configurations/archive_file.py:56]: do we need a `result()` like the others?

```suggestion
    ).result()
```

File: snippets/sdk/python/deployments/prerequisites.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    resource_client.resource_groups.begin_delete(GROUP_NAME).result()
+    print("Deleted resource group.\n")
+
+
+def main():
+    create()
+    delete()
+
+
+if __name__ == "__main__":
```
Comment (ryepup) [inline snippets/sdk/python/deployments/prerequisites.py:163]: is this file intended to be run as a script, or just in support of the other scripts?

File: snippets/sdk/python/certificates/create_or_update.py (hunk 6 above)
Comment (puneetsarna) [reply snippets/sdk/python/certificates/create_or_update.py:59]: that or a `constants.py` or something.

File: snippets/sdk/python/README.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+### Prerequisites
+
+1. Before we run the samples, we need to make sure we have setup the credentials. Follow the instructions in [register a new application using Azure portal](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal) to obtain `subscription id`,`client id`,`client secret`, and `application id`
+
+2. Store your credentials an environment variables.
+For example, in Linux-based OS, you can do
+
+```bash
+export AZURE_TENANT_ID="xxx"
```
[the remark this answers is not in the document]
Comment (agazeley) [reply snippets/sdk/python/README.md:23]: I can remove all this. I was going by what MSFT has been doing for their SDK snippets repo.

Example: https://github.com/Azure-Samples/azure-samples-python-management/blob/main/samples/storage/README.md

File: snippets/sdk/python/README.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+export AZURE_CLIENT_ID="xxx"
+export AZURE_CLIENT_SECRET="xxx"
+export AZURE_SUBSCRIPTION_ID="xxx"
+```
+
+### Installation
```
[the remark this answers is not in the document]
Comment (agazeley) [reply snippets/sdk/python/README.md:29]: I can remove all this. I was going by what MSFT has been doing for their SDK snippets repo.

Example: https://github.com/Azure-Samples/azure-samples-python-management/blob/main/samples/storage/README.md

File: snippets/sdk/python/deployments/prerequisites.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+
+    # Create user assigned managed identity
+    identity = msi_client.user_assigned_identities.create_or_update(
+        GROUP_NAME, MANAGED_IDENTITY_NAME, {"location": "eastus"}
+    )
+    print("Created a managed identity.\n")
+    return subnet, public_ip, identity
+
+
+def delete():
```
[the remark this answers is not in the document]
Comment (agazeley) [reply snippets/sdk/python/deployments/prerequisites.py:123]: We for sure can.I was going by what MSFT has been doing for their SDK snippets repo where the specifically create and delete each element to show you how to do all the operations you may want to perform.

Example:https://github.com/Azure-Samples/azure-samples-python-management/blob/main/samples/storage/manage_blob_container.py

File: snippets/sdk/python/deployments/create_or_update.py (hunk 2 above)
Comment (agazeley) [reply snippets/sdk/python/deployments/create_or_update.py:26]: This is to provide folks an example of the structure needed for a private IP. I can move this to a different snippet though like we do for arm-templates.

File: snippets/sdk/python/deployments/with_metrics.py (hunk 10 above)
Comment (agazeley) [reply snippets/sdk/python/deployments/with_metrics.py:1]: That works for me.

File: snippets/sdk/python/README.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+1. If you don't already have it, [install Python](https://www.python.org/downloads/).
+
+2. General recommendation for Python development is to use a Virtual Environment.
+    For more information, see <https://docs.python.org/3/tutorial/venv.html>
```
[the remark this answers is not in the document]
Comment (agazeley) [reply snippets/sdk/python/README.md:34]: I was unaware bare links was an issue? I do like the suggestion.

File: snippets/sdk/python/certificates/create_or_update.py (hunk 4 above)
Comment (agazeley) [reply snippets/sdk/python/certificates/create_or_update.py:74]: Sadly nope we have to use the magic UUIDs. I can try and find the SDK constant that refers to it. Everywhere I have seen role assignment API docs it has used these magic strings and pointed to the official docs for folks who are interested in more role assignment role IDs. I tried to document what that UUID was and where to find all of them consistently with what MSFT docs and SDK snippets use.

File: snippets/sdk/python/configurations/archive_file.py (hunk 11 above)
Comment (agazeley) [reply snippets/sdk/python/configurations/archive_file.py:14]: Should be documented in both snippets examples. I really wish we did not require the configuration name and it just defaulted to....default.

File: snippets/sdk/python/deployments/prerequisites.py (hunk 13 above)
Comment (agazeley) [reply snippets/sdk/python/deployments/prerequisites.py:163]: All these files are intended to be able to be run as scripts. The prereqs one is only broken up so that it can be run standalone or use parts of it in the deployment scripts.

File: snippets/sdk/python/deployments/create_or_update.py (hunk 1 above)
Comment (agazeley) [reply snippets/sdk/python/deployments/create_or_update.py:56]: Other than to provide an example of how to update a deployment. No.
